stub_uploader/metadata.py

```diff
@@ -1,19 +1,39 @@
+from __future__ import annotations
+
+import functools
+import graphlib
 import os
-from typing import Any, Dict, List, Optional
+import re
+from typing import Any, Dict, Iterator, Optional
 
+import requests
 import tomli
+from packaging.requirements import Requirement
+
+from .const import META, THIRD_PARTY_NAMESPACE, TYPES_PREFIX, UPLOADED_PATH
+
 
-from .const import META, THIRD_PARTY_NAMESPACE, TYPES_PREFIX
+class InvalidRequires(Exception):
+    pass
 
 
 class Metadata:
     def __init__(self, distribution: str, data: Dict[str, Any]):
-        self.upstream_distribution = distribution
+        assert not distribution.startswith(TYPES_PREFIX)
+        self._alleged_upstream_distribution = distribution
         self.data = data
 
+    @property
+    def upstream_distribution(self) -> Optional[str]:
+        # TODO: add a field to METADATA.toml if the stubs are for a package
+        # that does not exist on PyPI
+        if self._alleged_upstream_distribution == "gdb":
```

Comment:
This makes me think about another possible attack. What if someone creates a package `foo` and adds `types-foo` to typeshed. It can be a totally benign package. Then later they make a PR that makes `types-foo` a dependency of something commonly used, like `types-requests` (or maybe something less suspicious), e.g. because the library has some custom type that works as an argument for some `requests` function. Much later they add some malicious package `bad` as a dependency of `foo`, and submit a "trivial" PR adding it also as a dependency of `types-foo`. Now this malicious package `bad` will be installed for _a lot_ of users (executing arbitrary code during install), despite the fact that no-one may actually use `foo`.

Of course this is a tricky scenario, but someone may still try it, taking into account that potential impact will be very high. You may say that something similar is also possible for a regular runtime package, but the difference is that maintainers of runtime packages usually deal with only few dependencies, and adding another one is a really big deal, while typeshed maintainers will (hopefully) deal with hundreds or thousands packages each with their own deps.

So, I would say we will still need an explicit allowlist _in this repo_, in addition to the logic proposed in this PR.

Reply:
I think that can be dealt with in a similar way to the current logic for external dependencies, that is, in order to add `types-xyz` as a dependency of `types-requests`, `xyz` needs to be a dependency of `requests`.

That's probably more robust too, because whatever process we have for adding things to the allowlist will be used to add things to the allowlist. To spell things out:
1. Package foo exists, that depends on bar
2. Add types-foo, add dependency on bar
3. In order to make types-foo useful, we need bar, so we do some security review of bar and add it to the allowlist
4. bar adds dependency on evil
5. types-foo users are affected, this isn't good but we're okay with it, since those are assumed to be a subset of foo users
6. Someone sneaks past adding a dependency on types-foo to types-requests
7. types-requests users are affected, this is really bad

So adding a step 3 can help in practice, but we should prevent step 6. I can write this check, should be pretty easy to extend the existing logic.

Note that I'm not against adding an allowlist, but I'd like to think of it more as a fallback. I'd also want us to clearly document the criteria for additions to the allowlist.

Reply:
Okay, I wrote the logic for preventing step 6 locally, but it doesn't pass cleanly against existing packages. There are four failures, two of which are addressed by https://github.com/python/typeshed/pull/8792 and https://github.com/python/typeshed/pull/8793.

The other two are trickier, so would need to make that change later in another PR (might need new METADATA.toml fields). In the meantime I've added the allowlist as proposed, which should unblock this PR.

Reply:
> but it doesn't pass cleanly against existing packages

Yeah, I felt it may happen. Anyway, the PR looks good now.